### PR TITLE
Configurator rest

### PR DESCRIPTION
**Pull request by hiteshhapani**

### CODE REVIEW

File: gbpservice/neutron/nsf/configurator/api/root_controller.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from pecan import expose
+from v1 import controllers
+
+
+class RootController(object):
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/root_controller.py]: Does this file comes as part of pecan installation? If so, what is changed in this file from base version to what is checked in?

File: gbpservice/neutron/nsf/configurator/api/config.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Server Specific Configurations
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/config.py]: Does this file comes as part of pecan installation?

File: gbpservice/neutron/nsf/configurator/api/setup.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+try:
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/setup.py]: Does this file comes as part of pecan installation?

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/__init__.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+from pecan import expose
+from pecan import rest
+from wsmeext import pecan as wsme_pecan
+from wsme import types as wtypes
+
+import controller
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/__init__.py]: Imports should be in alphabetical order

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/__init__.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+
+class ControllerResolver(object):
+    """this class send parameter to controller class
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/__init__.py]: Class commenting should follow PEP compliance. Keep this outside the class.

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/__init__.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        "update_network_device_config")
+    create_network_service_config = controller.Controller(
+        "create_network_service_config")
+    delete_network_service_config = controller.Controller(
+        "delete_network_service_config")
+    update_network_service_config = controller.Controller(
+        "update_network_service_config")
+    get_notifications = controller.Controller("get_notifications")
+
+
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/__init__.py]: Add class comments

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/__init__.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+class V1Controller(object):
+
+    """ all request with nsf in query land here"""
+    nfp = ControllerResolver()
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/__init__.py]: Why is the class name ControllerResolver?

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/constants.py [hunk 8]
```diff
@@ -0,0 +1 @@
+TOPIC = "configurator"
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/constants.py]: If this constant is used in single location, it can be placed at the start of that file.

Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/constants.py]: Name the constant as CONFIGURATOR_RPC_TOPIC

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
+import json
+import subprocess
+
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py]: Organize the imports in alphabetical order otherwise this code will not be submittable.

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+from neutron.common import rpc as n_rpc
+from neutron.agent.common import config
+from oslo_config import cfg
+import oslo_messaging
+from pecan import expose, request
+from pecan import rest
+
+import constants
+
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py]: Add class and function comments.

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+
+class Controller(rest.RestController):
+
+    """controller class for handling all the curl request"""
+
+    def __init__(self, module_name):
+        self.host = subprocess.check_output('hostname', shell=True).rstrip()
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py]: Handle exceptions. Log errors.

Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py]: Any shell command that we run should be in a separate library. Or any OS specific code should be in separate library. When configurator moves to different OS, it will be easy to port the code.

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+        try:
+            return self._get_notifications()
+        except Exception as e:
+            return json.dumps({'err_msg': e.args})
+
+    @expose(method='POST', content_type='application/json')
+    def post(self, **body):
+        try:
+            body = None
+            if request.is_body_readable:
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py]: If the request body is not readable, why are we proceeding further to send the RPC call to configurator?

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+                body = request.json_body
+            if self.module_name == "create_network_device_config":
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py]: Can we construct the method name using getattr? It will be single line if we do this.

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+        self.rpcclient = RPCClient(topic=constants.TOPIC, host=self.host)
+        self.module_name = module_name
+        super(Controller, self).__init__()
+
+    @expose(method='GET', content_type='application/json')
+    def get(self):
+        try:
+            return self._get_notifications()
+        except Exception as e:
+            return json.dumps({'err_msg': e.args})
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py]: What is the error code that config agent proxy expects? Is this the format?

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+                raise Exception('Invalid Request')
+        except Exception as e:
+            return json.dumps({'err_msg': e.args})
+
+    @expose(method='PUT', content_type='application/json')
+    def put(self, **body):
+        try:
+            body = None
+            if request.is_body_readable:
+                body = request.json_body
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py]: Shouldn't we return if body is not readable?

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+        try:
+            return json.dumps(
+                self.rpcclient.get_notifications())
+        except Exception as e:
+            return json.dumps({'err_msg': e.args})
+
+    def _create_network_device_config(self, body):
+        request_data = body.get("request_data")
+        try:
+            return json.dumps(
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py]: What are we getting out as return value for this? What is the return value format followed for successful cases?

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+        except Exception as e:
+            return json.dumps({'err_msg': e.args})
+
+    def _update_network_service_config(self, body):
+        request_data = body.get("request_data")
+        try:
+            return json.dumps(
+                self.rpcclient.update_network_service_config(request_data))
+        except Exception as e:
+            return json.dumps({'err_msg': e.args})
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py]: Need to log in case of failures/exceptions

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+        return cctxt.cast(self, 'delete_network_device_config',
+                          request_data=request_data)
+
+    def delete_network_service_config(self, request_data):
+
+        cctxt = self.client.prepare(server=self.host)
+        return cctxt.cast(self, 'delete_network_service_config',
+                          request_data=request_data)
+
+    def to_dict(self):
```
Comment (njagadish) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py]: Why is this needed? We should avoid this method. Context class must be having this.

File: gbpservice/neutron/nsf/configurator/api/config.py (hunk 2 above)
Comment (hiteshhapani) [inline gbpservice/neutron/nsf/configurator/api/config.py]: yes, But we need to change path of root controller and module in this file according to our requirement.

File: gbpservice/neutron/nsf/configurator/api/root_controller.py (hunk 1 above)
Comment (hiteshhapani) [inline gbpservice/neutron/nsf/configurator/api/root_controller.py]: default pecan installation does not give this file. But pecan create [project_name] provides root.py file with sample get method. This file provides path to our controller.

File: gbpservice/neutron/nsf/configurator/api/setup.py (hunk 3 above)
Comment (hiteshhapani) [inline gbpservice/neutron/nsf/configurator/api/setup.py]: No. This also gets created  as a part of pecan create [project_name] command. But I assumed that we are not going to run that command and directly paste our directory so I added this with required modification.

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/__init__.py (hunk 4 above)
Comment (hiteshhapani) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/__init__.py]: I put it in alphabetical order but grouped it in standard, third party and local imports according to the guidelines.

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py (hunk 12 above)
Comment (hiteshhapani) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py]: If request body is not readable it wont proceed for RPC call. It will throw an error.

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py (hunk 13 above)
Comment (hiteshhapani) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py]: yes. I will change this

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py (hunk 16 above)
Comment (hiteshhapani) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py]: In case of get_notification we are getting dictionary that will be returned as it is. In all other cases we are using cast so if our cast call is successful it will return null.

File: gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py (hunk 18 above)
Comment (hiteshhapani) [inline gbpservice/neutron/nsf/configurator/api/v1/controllers/controller.py]: context is coming as a dictionary in request data. So anyway we need to add this method either here or in separate class. This method must be passed in call/cast.
